mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -0,0 +1,523 @@
+    fn r(&self) -> RwLockReadGuard<WatcherMut> { self.mutable.read().unwrap() }
+
+    #[inline]
+    fn apply_event(&self, event: WatcherEvent) {
+        match event {
+            WatcherEvent::Started => (),
+            WatcherEvent::StartFailed(err) => self.errors.lock().push(err),
+            WatcherEvent::TakerPaymentWaitConfirmFailed(err) => self.errors.lock().push(err),
+            WatcherEvent::TakerPaymentValidatedAndConfirmed => (),
+            WatcherEvent::TakerPaymentValidateFailed(err) => self.errors.lock().push(err),
+            WatcherEvent::TakerPaymentSpent(data) => {
+                self.w().taker_payment_spend = Some(data.transaction);
+                self.w().secret = data.secret;
+            },
+            WatcherEvent::TakerPaymentWaitForSpendFailed(err) => self.errors.lock().push(err),
+            WatcherEvent::MakerPaymentSpendFailed(err) => self.errors.lock().push(err),
+            WatcherEvent::MakerPaymentSpent(tx) => self.w().maker_payment_spend = Some(tx),
+            WatcherEvent::Finished => (),
+        }
+    }
+
+    async fn handle_command(
+        &self,
+        command: WatcherCommand,
+    ) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
+        match command {
+            WatcherCommand::Start => self.start().await,
+            WatcherCommand::ValidateTakerPayment => self.validate_taker_payment().await,
+            WatcherCommand::WaitForTakerPaymentSpend => self.wait_for_taker_payment_spend().await,
+            WatcherCommand::SpendMakerPayment => self.spend_maker_payment().await,
+            WatcherCommand::Finish => Ok((None, vec![WatcherEvent::Finished])),
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        uuid: Uuid,
+        ctx: MmArc,
+        maker_coin: MmCoinEnum,
+        taker_coin: MmCoinEnum,
+        data: TakerSwapWatcherData,
+    ) -> Self {
+        Watcher {
+            uuid,
+            ctx,
+            maker_coin,
+            taker_coin,
+            errors: PaMutex::new(Vec::new()),
+            mutable: RwLock::new(WatcherMut {
+                taker_payment_spend: None,
+                maker_payment_spend: None,
+                secret: H256Json::default(),
+            }),
+            data,
+        }
+    }
+
+    async fn start(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
+        Ok((Some(WatcherCommand::ValidateTakerPayment), vec![WatcherEvent::Started]))
+    }
+
+    // Do we need the exact same validation as the maker, or should we use a simpler validation process?
+    async fn validate_taker_payment(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
+        let wait_duration = (self.data.lock_duration * 4) / 5;
+        let wait_taker_payment = self.data.swap_started_at + wait_duration;
+        let confirmations = self.data.taker_payment_confirmations;
+
+        // Does the watcher have to wait for the confirmations like the maker does?
```

Comment:
I think it's enough to wait for one confirmation in watchers' case. We will also need to have a selection algorithm to prevent dozens of watchers tracking the same swap and causing excessive load on remote servers.

Regarding native mode - it might be not trivial to implement swap watchers for this case - it will require importing each swap P2SH addresses before the payment is sent. And I also wonder about daemons' performance in case of importing thousands of different addresses to local index. @caglaryucekaya That's also an item for further research checklist 🙂

Reply:
I also think that waiting is necessary, but it makes sense to reduce the number of confirmations to one. It will also reduce one field from the `TakerSwapWatcherData`.

Reply:
> Is it possible to arrange swaps without waiting for any confirmations?

I guess so since a transaction can be mined and spent in the same block (it just has to come before the transaction spending it), but it isn't safe to do that from the maker side I think, since a transaction that is unconfirmed might not be mined and the taker can claim the maker payment on the other chain since the secret was revealed. That's why it's left to the maker to decide how much confirmation he has to wait, so if he decides 0 confirmation it will be at his own risk.

> It looks like adding the confirmation number to the `TakerSwapWatcherData` is still necessary in case it is set to 0.

I think waiting for one confirmation in watchers' case as @artemii235 said is ok even if `taker_payment_confirmations` is 0, since watchers will wait for the maker spending the taker's payment on the next step anyways and we can assume that the maker will not spend the taker's payment before at least one confirmation, but we need @artemii235 to confirm this (no pun intended 😄 )

I am not even sure if `wait_for_confirmations` is needed at all in watchers case since `wait_for_tx_spend` will will wait for the maker's payment anyways after the maker confirms the taker payment, but maybe `wait_for_confirmations` is less resource intensive for the watchers and/or electrums. What do you think again @caglaryucekaya @artemii235 ?

Reply:
Hmm I was thinking that the point of watchers waiting for taker payment confirmations is to make sure that the taker payment actually exist. Otherwise attackers could flood the watchers with fake `TakerSwapWatcherData` and make them wait for long periods for maker payments that will never come. But actually the duration of `wait_for_confirmations` is only slightly shorter than the duration of `wait_for_tx_spend` so I'm also not sure if it's worth it. So let's wait for Artem's confirmation :smiley:

Reply:
@caglaryucekaya @shamardy Yeah, waiting for confirmation seems not required. But watchers should check if transactions exist because they might be dropped from mempool. Even if HTLC is not mined, it can be still spent by maker (if parties use 0 confirmations, though it's risky and only utilized for test purposes) or refunded.

Reply:
> @caglaryucekaya @shamardy Yeah, waiting for confirmation seems not required. But watchers should check if transactions exist because they might be dropped from mempool.

After this PR https://github.com/KomodoPlatform/atomicDEX-API/pull/1446 is merged, `wait_for_confirmations` will return an error straight away if the transaction is dropped from the mempool. So maybe we can keep `wait_for_confirmations` here after all, with confirmations equal to 1 if `taker_payment_confirmations` is not 0 like it's currently implemented after the fix. `taker_payment_confirmations` equal to zero should be allowed for test purposes as @artemii235 mentioned.